src/compiler/checker.ts

```diff
@@ -14860,6 +14860,14 @@ namespace ts {
                         target = removeTypesFromUnionOrIntersection(<UnionOrIntersectionType>target, matchingTypes);
                     }
                 }
+                else if (target.flags & TypeFlags.Substitution) {
+                    target = (target as SubstitutionType).typeVariable;
+                }
+                else if (target.flags & TypeFlags.IndexedAccess && (
+                    (<IndexedAccessType>target).objectType.flags & TypeFlags.Substitution ||
+                    (<IndexedAccessType>target).indexType.flags & TypeFlags.Substitution)) {
+                    target = getIndexedAccessType(getActualTypeVariable((<IndexedAccessType>target).objectType), getActualTypeVariable((<IndexedAccessType>target).indexType));

```

Comment:
What about a `Substitute<T[Substitute<K>]>`? Substitutes can in theory nest arbitrarily, provided they keep instantiating to type variables.

Reply:
I've moved the logic to `getActualTypeVariable` and made it recurse appropriately.